### PR TITLE
Exposing AttributeSet::copyAttributeValues via PointDataLeafNode

### PR DESCRIPTION
**Pull request by jbrd**
It would be useful to be able to use AttributeSet::copyAttributeValues on an existing PointDataLeafNode. Because the PointDataLeafNode only exposes a const AttributeSet, and cannot be constructed from an AttributeSet, there was no way of really using this method in practise. This was necessary in order to port the PointSort algorithm across to the latest public sources...

### CODE REVIEW

**Comment by danrbailey** [comment]
Thanks. I'll have to think about this a bit more. The copyAttributeValues method is so specific to the sorting, I really don't see it being used anywhere else, so not sure this is the right approach here (I was never completely happy with the original implementation).

Perhaps a helper sort intermediate class that has the PointDataGrid as a friend to get access to this data or something along these lines.

**Comment by jbrd** [comment]
Thanks for the feedback Dan. 

IMO the public API should expose everything we need in order to implement these algorithms, without us having to rely on friend-tactics. If we find ourselves having to internally expose pieces of data for these algorithms, then it could be taken as a sign that our public API isn't fully featured enough to be useful in practise.

I agree this is very use case specific and implementing this outside of the AttributeSet / PointDataLeafNode sounds like the best way forward. However, in order to do this we would need some way of copying a value from one AttributeArray to another without knowing what the underlying value-type is. At the moment, you can only obtain a TypedAttributeArray or an AttributeHandle from a PointDataLeafNode, and both require you to define the value type at compile-time. Would you have any objections to adding an attributeArray(size_t index) accessor to the PointDataLeafNode, which would return a reference to the underlying AttributeArray? 

With this, we should then be able to implement this functionality using the get / set methods on the AttributeArray itself, and we wouldn't have to maintain a large switch of value types. Furthermore, we could also remove copyAttributeValues on the AttributeSet - there is simply no way to use this in practise, given that its effects can never be seen by a PointDataLeafNode...

**Comment by jbrd** [comment]
Further to my previous comments, I've pushed up a set of changes with what I am proposing here:

* Exposing a typeless attributeArray accessor on PointDataLeafNode
* To go alongside this, I thought it might also be useful to provide a typeless hasAttribute method. This is separate and easy to revert if you'd prefer not to add this
* Removed copyAttributeValues from AttributeSet. Again, easy to revert if you wish to keep this, but as there is no way of benefiting from its functionality on a point data grid, I can't think of a good reason to keep it

I have an implementation of the PointSort algorithm that relies purely on these changes (where the functionality that was provided by copyAttributeValues now lives inside the PortSort).

File: openvdb_points/tools/PointDataGrid.h [hunk 1]
```diff
@@ -299,6 +299,22 @@ class PointDataLeafNode : public tree::LeafNode<T, Log2Dim> {
...
     /// than hasAttribute(const Name&) as it avoids a map lookup.
     /// @param pos    Index of the attribute
+    bool hasAttribute(const size_t pos) const
+    {
+        return pos < mAttributeSet->size();
+    }
+
+    /// @brief Returns whether an attribute exists.
+    /// @param attributeName    Name of the attribute
+    bool hasAttribute(const Name& attributeName) const
```
Comment (danrbailey) [inline openvdb_points/tools/PointDataGrid.h]: I like this change, think we should also change the original methods to hasTypedAttribute() to distinguish them.

File: openvdb_points/tools/AttributeSet.h [hunk 2]
```diff
@@ -155,9 +155,6 @@ class AttributeSet
     ///         shared with anyone else.
     void makeUnique(size_t pos);
 
-    /// Copy attribute values for a target index @a n from another @a attributeSet for index @a source
```
Comment (danrbailey) [inline openvdb_points/tools/AttributeSet.h]: I agree with removing this method, there's currently no functionality that uses it.

File: openvdb_points/tools/PointDataGrid.h [hunk 3]
```diff
@@ -328,6 +344,32 @@ class PointDataLeafNode : public tree::LeafNode<T, Log2Dim> {
         mAttributeSet->appendAttribute(attribute, expected, replacement);
     }
 
+    AttributeArray& attributeArray(const size_t pos)
```
Comment (danrbailey) [inline openvdb_points/tools/PointDataGrid.h]: I still don't really see any reason to expose this at the moment with the exception of the sorting routine, so I still feel like there might be a better way of achieving this. I can take a look at the changes you've made to the sort tomorrow and see if we can find another way.

**Comment by jbrd** [comment]
Here is another example that demonstrates why access to the underlying AttributeArray objects would be useful.

I was recently debugging a PointDataGrid and suspected that the voxel offset values were not consistent with the attribute arrays on their respective leaves. So I wrote a small sanity check function that:

* For each leaf
  * For each voxel
    * Obtain the voxel offset value
    * For each attribute array on the leaf
      * Check that the voxel offset value was within the bounds of the attribute array

Having access to the base AttributeArrays made this trivial to implement - I just obtain each AttributeArray by index and call its 'size' method.

If our only option is to call ```typedAttributeArray<T>```, then I would have had to do something like this:

* For each leaf
  * For each voxel
    * Obtain the voxel offset value
    * For each attribute array on the leaf
      * Is the attribute array of type float + null attribute codec?
        * Call ```typedAttributeArray<Float, NullAttributeCodec>``` and call its 'size' method
      * Is the attribute array of type int + null attribute codec?
        * Call ```typedAttributeArray<Int, NullAttributeCodec>``` and call its 'size' method
      * Is the attribute array of type double + null attribute codec?
        * Call ```typedAttributeArray<double, NullAttributeCodec>``` and call its 'size' method
      * repeat for all combinations of value type and codec, for which there are many!

In short, there are methods on the base class that are useful, in cases where you don't know in advance the value type and codec of every attribute array.

**Comment by jbrd** [comment]
I have now separated this out into three separate pull requests...
